README.md
# l10n-maven-plugin [![Build Status](https://travis-ci.org/rquinio/l10n-maven-plugin.svg)](https://travis-ci.org/rquinio/l10n-maven-plugin) [![Coverage Status](https://coveralls.io/repos/rquinio/l10n-maven-plugin/badge.svg)](https://coveralls.io/r/rquinio/l10n-maven-plugin) [![Maven Central](https://maven-badges.herokuapp.com/maven-central/com.googlecode.l10n-maven-plugin/l10n-maven-plugin/badge.svg)](https://maven-badges.herokuapp.com/maven-central/com.googlecode.l10n-maven-plugin/l10n-maven-plugin)

> A Maven plugin to validate localization resources in Java properties files 

## Description

The localization process of an application often involves non technical people. Some translated resources may contain issues linked to formatting, encoding, or the context where the resource is used.

l10-maven-plugin can validate a set of [properties files](http://en.wikipedia.org/wiki/.properties) (mainly used in Java [ResourceBundle](http://docs.oracle.com/javase/1.5.0/docs/api/java/util/ResourceBundle.html)). It aims to either:
  * Detect invalid l10n properties resources at build time, typically before the webapp is packaged, or when extracting properties files from a localization tool/database (**[l10n:validate goal](https://github.com/rquinio/l10n-maven-plugin/wiki/l10n:validate goal)**).
  * Build a Maven site report listing violations found on properties (**[l10n:report goal](https://github.com/rquinio/l10n-maven-plugin/wiki/l10n:report goal)**), see a [sample report](https://cdn.rawgit.com/wiki/rquinio/l10n-maven-plugin/SampleReport.html).

The plugin was initially developed for some webapps translated in 7 languages with several thousands of properties across 5 resource bundles. It allowed to prevent buggy translations from slipping into production and reduced the cost of fixing them by detecting issues earlier.

It aims to be a pragmatic solution when having legacy constraints or needing fast & easy tool, yet probably not the ideal solution in terms of usability:
  * Some checks could rather be performed as soon as translator uploads/inputs the translation in the localization tool, to provide instant feedback.
  * Encoding/escaping should rather be managed by the code using the resources, based on the context they are being used (server side, client side, ...) rather than in the resources themselves.

## Usage

See [Usage](https://github.com/rquinio/l10n-maven-plugin/wiki/Usage) page for plugin goals and detailed configuration.

Plugin is available on [Maven Central](http://search.maven.org/#search|ga|1|g%3A%22com.googlecode.l10n-maven-plugin%22), through [Sonatype OSS hosting](https://oss.sonatype.org).

```xml
<plugin>
  <groupId>com.googlecode.l10n-maven-plugin</groupId>
  <artifactId>l10n-maven-plugin</artifactId>
  <version>1.8</version>
</plugin>
```

## Requirements

The following specifies the _minimum_ requirements to run this Maven plugin:

| Library | Min version | Notes |
|--------|------------|------|
| **Maven** | **2.2.1** | Might work on previous Maven 2 versions, but not tested. |
| **JDK** | **1.5** |  |


## Validation

See the list of [validators](https://github.com/rquinio/l10n-maven-plugin/wiki/Validators) and associated checks.

## References

  * [java.util.Properties#load](http://docs.oracle.com/javase/1.5.0/docs/api/java/util/Properties.html#load%28java.io.InputStream%29) Javadoc
  * [java.text.MessageFormat](http://docs.oracle.com/javase/1.5.0/docs/api/java/text/MessageFormat.html) and [java.util.Formatter](http://docs.oracle.com/javase/1.5.0/docs/api/java/util/Formatter.html) Javadoc.
  * [Javascript special characters](http://www.w3schools.com/js/js_special_characters.asp)
  * [java.util.Pattern](http://docs.oracle.com/javase/1.5.0/docs/api/java/util/regex/Pattern.html) Javadoc
  * [JSON specification](http://json.org/)


ChangeLog.txt
1.9 Next
---------------
#17 Make the search for bundles in propertyDir recursive. In order to easily differentiate bundles with identical names but in different sub-folders, their file and bundle names are fully qualified to reflect their relative paths inside propertyDir.

1.8 30-Nov-2014
---------------
#8 Change absolute URI in xhtml1 schemas to relative URI, in order to avoid re-downloading xml.xsd each time.
#11 Make plugin threadsafe for maven 3.x parallel execution
#12 Support for property substitution (i.e. property = ${anotherProperty}) in property values.

1.7 15-Nov-2013
---------------
#1 Support for String.format parameters (C-stytle formatting)
#4 Support for multi-line property value with parameters and quotes
#5 Support for multiple resource bundles per project
#7 Support for detecting duplicated keys in property file

1.6 07-Jan-2013
---------------
Added French translation for the plugin
Fixed a bug with a simple quote following a double quote in the HTML validation
Addition of the acceptance of LI elements at the root of an HTML resource (without a UL or OL encompassing element)

1.5 23-Sep-2012
---------------
Added generation of CSV files with keys for each severity (ERROR/WARN/INFO).
Added a summary to the report, with links to sections
Added logging of expected html/url key pattern for default validator
Refactored ValidateMojo logic by splitting the class into 6 orchestrators
Fixed strange bug where javascript validator would not log errors into the console
Fixed HTML tag coherence validator to ignore siblings order
Upgraded xhtml5 schema to XMLmind 1.0.3

1.4 26-May-2012 New validators
---------------
Refactored validators architecture and model, with notion of scope (single property value, translations of a property, whole properties file, whole bundle of properties file, ...)
Added identical translation validator, to info identically translated resources, or warn if translations are mostly identical but a few differ.
Added Trailing whitespace validator for all resources.
Added SpellCheck validator, based on Jazzy Spellchecker and using locale from properties file name. Location of dictionaries has to be specified via plugin configuration.
Added HTML tag coherence validator, to warn if html tags used are different between translated values of a resource.
Modified URL validator to detect HTML imported urls with hardcoded http protocol (*.js, *.css, *.png, ...) that would cause browser mixed content warnings.
Modified HTML validator to ignore HTML5 data-* custom attributes, as this cannot easily be expressed in an XML schema.
Added threshold (30) for number of info/warn issues logged per type.
Added skip plugin configuration property.
Added plugin integration tests using maven-invoker-plugin.
Added log of issues summary per severity & type, after validation.

1.3 07-Apr-2012 Customization
---------------
Added custom pattern validators.
Added XHTML validation schema configuration, with pre-defined xhtml1-transitional (default), xhtml1-strict or xhtml5 (based on http://www.xmlmind.com/xhtml5_resources.shtml 1.0.1).
Added configuration to switch from double quoted js resources (default) to single quoted js resources validation. 
Modified URL validation regex to use more precise ESAPI pattern.
Increased number of replaced formatting parameters in resource before URL/XHTML validation from 9 to 19.
Fixed report generation failure when no issues found.

1.2.1 14-Feb-2012 Java 5 compatibility patch
---------------
Fixed Java 5 compatibility issue introduced in 1.2.

1.2 14-Feb-2012 Reporting goal
---------------
Added report goal that binds to site phase;
Added parametric replacement checks (single quote escaping, parameters coherency).
Added support for javascript escaping of double quotes in js resources (\\").
Added javascript unescaping before HTML validation for js resources.
Added missing translation checks.
Refactored code architecture with validators.
Changed logging to be more consistent.

1.1 10-Feb-2012
---------------
Added HTML unescaping before URL validation.
Added Text key validation (no HTML/URL).
Changed Other resources validation to warning instead of error.

1.0 07-Feb-2012 Initial release
---------------
Validate goal that binds to test phase.


src/main/java/com/googlecode/l10nmavenplugin/ValidateMojo.java
/*******************************************************************************
 * Copyright (c) 2012 Romain Quinio (http://code.google.com/p/l10n-maven-plugin)
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 ******************************************************************************/
package com.googlecode.l10nmavenplugin;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

import org.apache.maven.plugin.AbstractMojo;
import org.apache.maven.plugin.MojoExecutionException;
import org.apache.maven.plugin.MojoFailureException;
import org.apache.maven.plugins.annotations.LifecyclePhase;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;

import com.googlecode.l10nmavenplugin.format.Formatter;
import com.googlecode.l10nmavenplugin.format.InnerResourcesFormatter;
import com.googlecode.l10nmavenplugin.log.L10nValidatorLogger;
import com.googlecode.l10nmavenplugin.model.L10nReportItem;
import com.googlecode.l10nmavenplugin.model.Property;
import com.googlecode.l10nmavenplugin.model.PropertyFamily;
import com.googlecode.l10nmavenplugin.validators.L10nValidationException;
import com.googlecode.l10nmavenplugin.validators.L10nValidator;
import com.googlecode.l10nmavenplugin.validators.family.HtmlTagCoherenceValidator;
import com.googlecode.l10nmavenplugin.validators.family.IdenticalTranslationValidator;
import com.googlecode.l10nmavenplugin.validators.family.MissingTranslationValidator;
import com.googlecode.l10nmavenplugin.validators.family.ParametricCoherenceValidator;
import com.googlecode.l10nmavenplugin.validators.file.DuplicateKeysValidator;
import com.googlecode.l10nmavenplugin.validators.orchestrator.DirectoryValidator;
import com.googlecode.l10nmavenplugin.validators.orchestrator.PropertiesFamilyValidator;
import com.googlecode.l10nmavenplugin.validators.orchestrator.PropertyFamilyValidator;
import com.googlecode.l10nmavenplugin.validators.orchestrator.PropertyValidator;
import com.googlecode.l10nmavenplugin.validators.property.DefaultValidator;
import com.googlecode.l10nmavenplugin.validators.property.HtmlValidator;
import com.googlecode.l10nmavenplugin.validators.property.JsValidator;
import com.googlecode.l10nmavenplugin.validators.property.PatternValidator;
import com.googlecode.l10nmavenplugin.validators.property.PlainTextValidator;
import com.googlecode.l10nmavenplugin.validators.property.SpellCheckValidator;
import com.googlecode.l10nmavenplugin.validators.property.TrailingWhitespaceValidator;
import com.googlecode.l10nmavenplugin.validators.property.UrlValidator;
import com.googlecode.l10nmavenplugin.validators.property.format.CStyleFormattingValidator;
import com.googlecode.l10nmavenplugin.validators.property.format.FormattingValidator;
import com.googlecode.l10nmavenplugin.validators.property.format.InnerResourcesFormattingValidator;
import com.googlecode.l10nmavenplugin.validators.property.format.MessageFormatFormattingValidator;

/**
 * Validate a set of l10n {@link Properties} files against:
 * 
 * <ul>
 * <li>Missing javascript escaping for resources evaluated client side</li>
 * <li>Bad escaping for {@link java.text.MessageFormat} in case of parametric replacement</li>
 * <li>Invalid XHTML 1.0 transitional</li>
 * <li>Malformed absolute URLs</li>
 * <li>Plain text resources containing HTML/URL</li>
 * </ul>
 * 
 * In case multiple checks are performed on a resource (ex: client side resource with parameters), the order above applies.
 * 
 * The syntax of properties file itself is not checked, but it relies on loading them successfully as {@link Properties} .
 * 
 * 
 * @note References for escape sequences and special characters:
 *       <ul>
 *       <li>java.util.Properties: {@link java.util.Properties#load}</li>
 *       <li>java.text.MessageFormat: {@link java.text.MessageFormat}</li>
 *       <li>java.util.Formatter {@link java.util.Formatter}</li>
 *       <li>Javascript: {@link http://www.w3schools.com/js/js_special_characters.asp}</li>
 *       <li>JSON {@link http://json.org/}</li>
 *       <li>XHTML: {@link http://www.w3schools.com/tags/ref_entities.asp}</li>
 *       <li>URL: {@link http://www.w3schools.com/tags/ref_urlencode.asp}</li>
 *       </ul>
 *       Extra references for development:
 *       <ul>
 *       <li>java.util.regex.Pattern: {@link java.util.regex.Pattern}</li>
 *       <li>Java String: {@link http://java.sun.com/docs/books/jls/second_edition/html/lexical.doc.html#101089}</li>
 *       </ul>
 * 
 * @since 1.0
 * @author romain.quinio
 * 
 */
@Mojo(name = "validate", defaultPhase = LifecyclePhase.TEST, threadSafe = true)
public class ValidateMojo extends AbstractMojo implements L10nValidationConfiguration {

  public static final String MESSAGE_FORMAT_FORMATTER = "messageFormat";

  public static final String C_STYLE_FORMATTER = "C-style";

  /**
   * Directory containing properties file to check.
   * <p>
   * The directory is scanned recursively.
   * 
   * @since 1.0
   */
  @Parameter(defaultValue = "src/main/resources")
  private File propertyDir;

  /**
   * Keys excluded from validation. Default is none.
   * 
   * @since 1.0
   */
  @Parameter
  private String[] excludedKeys = new String[] {};

  /**
   * Make validation failure not blocking the build
   * 
   * @since 1.0
   */
  @Parameter(defaultValue = "false")
  private boolean ignoreFailure = false;

  /**
   * List of keys to match as text resources used from js. Default is ".js.".
   * 
   * @since 1.0
   */
  @Parameter
  private String[] jsKeys = new String[] { ".js." };

  /**
   * Declares how the client side resources are loaded in javascript:
   * <ul>
   * <li>double quoted: var jsResource = "<fmt:message key='jsKey' />"</li>
   * <li>single quoted: var jsResource = '<fmt:message key="jsKey" />'</li>
   * <ul>
   * 
   * Default value is true (double quoted), which complies with JSON format.
   * 
   * @since 1.3
   */
  @Parameter(defaultValue = "true")
  private boolean jsDoubleQuoted = true;

  /**
   * List of keys to match as url resources. Default is ".url.".
   * 
   * @since 1.0
   */
  @Parameter
  private String[] urlKeys = new String[] { ".url." };

  /**
   * List of keys to match as html text resources. Default is ".text.".
   * 
   * @since 1.0
   */
  @Parameter
  private String[] htmlKeys = new String[] { ".text." };

  /**
   * XML Schema to use for html resource validation. Default value is to use XHTML1 transitional.
   * 
   * @since 1.3
   */
  @Parameter(defaultValue = "xhtml1-transitional.xsd")
  private File xhtmlSchema;

  /**
   * List of keys to match as non-html text resources. Default is ".title.".
   * 
   * @since 1.1
   */
  @Parameter
  private String[] textKeys = new String[] { ".title." };

  /**
   * Custom validation patterns.
   * 
   * @since 1.3
   */
  @Parameter
  private CustomPattern[] customPatterns = new CustomPattern[] {};

  /**
   * Directory containing dictionaries for SpellCheck validation.
   * 
   * Defaults to the value of parameter propertyDir.
   * 
   * @since 1.4
   */
  @Parameter
  private File dictionaryDir;

  /**
   * Flag allowing to skip plugin exceution for a particular build.
   * 
   * This makes the plugin more controllable from profiles.
   * 
   * @since 1.4
   */
  @Parameter(defaultValue = "${l10n.skip}")
  private boolean skip;

  /**
   * Base directory where all reports are written to.
   * 
   * @since 1.5
   */
  @Parameter(defaultValue = "${project.build.directory}/l10n-reports")
  private File reportsDir;

  /**
   * 
   * Type of {@link Formatter} used for parametric replacement.
   * 
   * Allowed values: {@link #MESSAGE_FORMAT_FORMATTER} and {@link #C_STYLE_FORMATTER}
   * 
   * @since 1.7
   */
  @Parameter(defaultValue = MESSAGE_FORMAT_FORMATTER)
  private String formatter;

  /**
   * Regular expression to use to match inner resources inside a resource value.
   * 
   * @since 1.8
   */
  @Parameter
  private String innerResourceRegex;

  private L10nValidator<File> directoryValidator;

  private L10nValidatorLogger logger;

  /**
   * Default constructor for plugin execution
   * 
   * Can't initialize validators here, because Mojo configuration is injected after constructor has returned
   * 
   */
  public ValidateMojo() {
  }

  /**
   * Initialize from another configured Mojo
   * 
   */
  public ValidateMojo(L10nValidationConfiguration configuration) {
    // Propagate configuration
    setPropertyDir(configuration.getPropertyDir());
    setHtmlKeys(configuration.getHtmlKeys());
    setXhtmlSchema(configuration.getXhtmlSchema());
    setJsKeys(configuration.getJsKeys());
    setJsDoubleQuoted(configuration.getJsDoubleQuoted());
    setTextKeys(configuration.getTextKeys());
    setUrlKeys(configuration.getUrlKeys());
    setCustomPatterns(configuration.getCustomPatterns());
    setExcludedKeys(configuration.getExcludedKeys());
    setDictionaryDir(configuration.getDictionaryDir());
    setReportsDir(configuration.getReportsDir());
    setFormatter(configuration.getFormatter());
    setInnerResourceRegex(configuration.getInnerResourceRegex());

    // Not useful, but for the sake of completeness
    setIgnoreFailure(configuration.getIgnoreFailure());
    setSkip(configuration.getSkip());
  }

  /**
   * Initialize validators from plugin configuration
   * 
   * TODO Should use a DI container to do the wirering
   * 
   */
  protected void initialize() {
    logger = new L10nValidatorLogger(getLog());
    getLog().info("Initializing l10n validators...");

    if (dictionaryDir == null) {
      // Default to propertyDir
      dictionaryDir = propertyDir;
    }
    L10nValidator<Property> spellCheckValidator = new SpellCheckValidator(logger, dictionaryDir);

    FormattingValidator formattingValidator;
    if (C_STYLE_FORMATTER.equalsIgnoreCase(formatter)) {
      formattingValidator = new CStyleFormattingValidator(logger);
    } else {
      formattingValidator = new MessageFormatFormattingValidator(logger);
    }
    Formatter formatter = formattingValidator.getFormatter();

    InnerResourcesFormattingValidator innerResourcesValidator = new InnerResourcesFormattingValidator(logger,
        innerResourceRegex);
    InnerResourcesFormatter innerResourceFormatter = innerResourcesValidator.getFormatter();

    L10nValidator<Property> htmlValidator;
    if (xhtmlSchema != null) {
      htmlValidator = new HtmlValidator(xhtmlSchema, logger, spellCheckValidator, htmlKeys, formatter,
          innerResourceFormatter);
    } else {
      htmlValidator = new HtmlValidator(logger, spellCheckValidator, htmlKeys, formatter, innerResourceFormatter);
    }

    L10nValidator<Property> jsValidator = new JsValidator(jsDoubleQuoted, htmlValidator, logger, jsKeys);
    L10nValidator<Property> urlValidator = new UrlValidator(logger, urlKeys, formatter);
    L10nValidator<Property> plainTextValidator = new PlainTextValidator(logger, spellCheckValidator, textKeys);
    L10nValidator<Property> defaultValidator = new DefaultValidator(logger, htmlKeys, urlKeys);

    L10nValidator<Property> trailingWhitespaceValidator = new TrailingWhitespaceValidator(logger);
    L10nValidator<PropertyFamily> missingTranslationValidator = new MissingTranslationValidator(logger);
    L10nValidator<PropertyFamily> parametricCoherenceValidator = new ParametricCoherenceValidator(logger, formatter);
    L10nValidator<PropertyFamily> identicalTranslationValidator = new IdenticalTranslationValidator(logger);
    L10nValidator<PropertyFamily> htmlTagCoherenceValidator = new HtmlTagCoherenceValidator(logger, htmlKeys);

    // L10nValidator<PropertiesFamily> duplicationValidator = new DuplicationValidator(logger);

    L10nValidator<Property>[] patternValidators = null;
    if (customPatterns != null) { // Initialize custom pattern validators
      patternValidators = new PatternValidator[customPatterns.length];
      for (int i = 0; i < customPatterns.length; i++) {
        CustomPattern pattern = customPatterns[i];
        patternValidators[i] = new PatternValidator(logger, pattern);
      }
    }

    PropertyValidator propertyValidator = new PropertyValidator(logger, excludedKeys);
    propertyValidator.setPatternValidators(patternValidators);
    propertyValidator.setDefaultValidator(defaultValidator);
    propertyValidator.setHtmlValidator(htmlValidator);
    propertyValidator.setJsValidator(jsValidator);
    propertyValidator.setParametricMessageValidator(formattingValidator);
    propertyValidator.setPlainTextValidator(plainTextValidator);
    propertyValidator.setTrailingWhitespaceValidator(trailingWhitespaceValidator);
    propertyValidator.setUrlValidator(urlValidator);
    propertyValidator.setInnerResourcesValidator(innerResourcesValidator);

    PropertyFamilyValidator propertyFamilyValidator = new PropertyFamilyValidator(logger, excludedKeys);
    propertyFamilyValidator.setHtmlTagCoherenceValidator(htmlTagCoherenceValidator);
    propertyFamilyValidator.setIdenticalTranslationValidator(identicalTranslationValidator);
    propertyFamilyValidator.setMissingTranslationValidator(missingTranslationValidator);
    propertyFamilyValidator.setParametricCoherenceValidator(parametricCoherenceValidator);
    propertyFamilyValidator.setPropertyValidator(propertyValidator);

    PropertiesFamilyValidator propertiesFamilyValidator = new PropertiesFamilyValidator(logger, reportsDir,
        propertyFamilyValidator);
    L10nValidator<File> duplicateKeysValidator = new DuplicateKeysValidator(logger, propertyDir);
    directoryValidator = new DirectoryValidator(logger, propertiesFamilyValidator, duplicateKeysValidator);
  }

  /**
   * Plugin entry point for validate goal.
   * 
   * @throws MojoExecutionException
   *           in case of unexpected exception during plugin execution
   * @throws MojoFailureException
   *           in case validation detected errors and ignoreFailure is false
   */
  public void execute() throws MojoExecutionException, MojoFailureException {
    if (!skip) {
      initialize();
      executeInternal();
    } else {
      getLog().info("Skipping plugin execution, as per configuration.");
    }
  }

  /**
   * Plugin entry point for unit testing to allow re-use of a single initialized Mojo instance, for perf reasons.
   * 
   * @throws MojoExecutionException
   *           in case of unexpected exception during plugin execution
   * @throws MojoFailureException
   *           in case validation detected errors and ignoreFailure is false
   */
  protected void executeInternal() throws MojoExecutionException, MojoFailureException {
    List<L10nReportItem> reportItems = new ArrayList<L10nReportItem>();

    int nbErrors = validate(propertyDir, reportItems);

    if (nbErrors > 0) {
      if (ignoreFailure) {
        getLog().error("Validation has failed with " + nbErrors + " errors.");
        getLog().info("Ignoring failure as ignoreFailure is true.");
      } else {
        throw new MojoFailureException("Validation has failed with " + nbErrors + " errors.");
      }
    } else {
      getLog().info("Validation was successful.");
    }
  }

  /**
   * Validation logic entry point used both by {@link ValidateMojo} and {@link ReportMojo}
   * 
   * @param directory
   *          the folder containing .properties files to validate
   * @param reportItems
   *          list to update with validation errors/warn/info items
   * @return number of validation errors
   * @throws MojoExecutionException
   */
  protected int validate(File directory, List<L10nReportItem> reportItems) throws MojoExecutionException {
    int nbErrors = 0;

    try {
      nbErrors = directoryValidator.validate(directory, reportItems);

    } catch (L10nValidationException e) {
      throw new MojoExecutionException(
          "An unexpected exception has occurred while validating properties under directory "
              + propertyDir.getAbsolutePath(), e);
    }
    return nbErrors;
  }

  public void setDirectoryValidator(L10nValidator<File> directoryValidator) {
    this.directoryValidator = directoryValidator;
  }

  public void setPropertyDir(File propertyDir) {
    this.propertyDir = propertyDir;
  }

  public void setExcludedKeys(String[] excludedKeys) {
    this.excludedKeys = excludedKeys;
  }

  public void setIgnoreFailure(boolean ignoreFailure) {
    this.ignoreFailure = ignoreFailure;
  }

  public void setJsKeys(String[] jsKeys) {
    this.jsKeys = jsKeys;
  }

  public void setUrlKeys(String[] urlKeys) {
    this.urlKeys = urlKeys;
  }

  public void setHtmlKeys(String[] htmlKeys) {
    this.htmlKeys = htmlKeys;
  }

  public void setTextKeys(String[] textKeys) {
    this.textKeys = textKeys;
  }

  public void setCustomPatterns(CustomPattern[] customPatterns) {
    this.customPatterns = customPatterns;
  }

  public void setJsDoubleQuoted(boolean jsDoubleQuoted) {
    this.jsDoubleQuoted = jsDoubleQuoted;
  }

  public void setXhtmlSchema(File xhtmlSchema) {
    this.xhtmlSchema = xhtmlSchema;
  }

  public void setDictionaryDir(File dictionaryDir) {
    this.dictionaryDir = dictionaryDir;
  }

  public File getPropertyDir() {
    return propertyDir;
  }

  public String[] getExcludedKeys() {
    return excludedKeys;
  }

  public boolean getIgnoreFailure() {
    return ignoreFailure;
  }

  public String[] getJsKeys() {
    return jsKeys;
  }

  public boolean getJsDoubleQuoted() {
    return jsDoubleQuoted;
  }

  public String[] getUrlKeys() {
    return urlKeys;
  }

  public String[] getHtmlKeys() {
    return htmlKeys;
  }

  public File getXhtmlSchema() {
    return xhtmlSchema;
  }

  public String[] getTextKeys() {
    return textKeys;
  }

  public CustomPattern[] getCustomPatterns() {
    return customPatterns;
  }

  public File getDictionaryDir() {
    return dictionaryDir;
  }

  public void setSkip(boolean skip) {
    this.skip = skip;
  }

  public boolean getSkip() {
    return skip;
  }

  public File getReportsDir() {
    return reportsDir;
  }

  public void setReportsDir(File reportsDir) {
    this.reportsDir = reportsDir;
  }

  public String getFormatter() {
    return formatter;
  }

  public void setFormatter(String formatter) {
    this.formatter = formatter;
  }

  public String getInnerResourceRegex() {
    return innerResourceRegex;
  }

  public void setInnerResourceRegex(String innerResourceRegex) {
    this.innerResourceRegex = innerResourceRegex;
  }

}
